sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -445,6 +472,10 @@ public BundleFinalizer getBundleFinalizer() {
                 throw new RuntimeException(e);
               }
             });
+    for (BeamFnDataOutboundAggregator aggregator :
+        bundleProcessor.getOutboundAggregators().values()) {
+      aggregator.startFlushThread();
+    }
```

Comment:
```suggestion
    for (BeamFnDataOutboundAggregator aggregator :
        bundleProcessor.getOutboundAggregators().values()) {
      aggregator.startFlushThread();
    }
```

Please move this to `BundleProcessor#finish`.

You should also add a BeamFnDataOutboundAggregator#discard method that cleans up the internal thread that it created and invoke it from `BundleProcessor#discard`.